File: articles/mathematical-biology-logic-living-systems/python/predator_prey.py
"""
Predator-prey workflows.

Run:
    python python/predator_prey.py
"""

from pathlib import Path

import numpy as np
import pandas as pd

from math_biology_core import simulate_predator_prey


ARTICLE_DIR = Path(__file__).resolve().parents[1]
PARAM_PATH = ARTICLE_DIR / "data" / "predator_prey_parameters.csv"


def main() -> None:
    scenarios = pd.read_csv(PARAM_PATH)

    rows = []

    for _, row in scenarios.iterrows():
        sim = simulate_predator_prey(
            prey0=row["prey0"],
            predator0=row["predator0"],
            alpha=row["alpha"],
            beta=row["beta"],
            delta=row["delta"],
            gamma=row["gamma"],
            time_end=row["time_end"],
            dt=row["dt"],
        )

        rows.append(
            {
                "scenario": row["scenario"],
                "final_prey": sim["prey"][-1],
                "final_predator": sim["predator"][-1],
                "max_prey": np.max(sim["prey"]),
                "max_predator": np.max(sim["predator"]),
                "mean_prey": np.mean(sim["prey"]),
                "mean_predator": np.mean(sim["predator"]),
            }
        )

    print(pd.DataFrame(rows).round(4).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/mathematical-biology-logic-living-systems/r/predator_prey.R
# Predator-prey workflow in R.

param_path <- file.path("data", "predator_prey_parameters.csv")
if (!file.exists(param_path)) {
  param_path <- file.path("..", "data", "predator_prey_parameters.csv")
}

simulate_predator_prey <- function(prey0, predator0, alpha, beta, delta, gamma, time_end, dt) {
  time <- seq(0, time_end, by = dt)
  prey <- numeric(length(time))
  predator <- numeric(length(time))

  prey[1] <- prey0
  predator[1] <- predator0

  for (i in 2:length(time)) {
    dprey <- alpha * prey[i - 1] - beta * prey[i - 1] * predator[i - 1]
    dpredator <- delta * prey[i - 1] * predator[i - 1] - gamma * predator[i - 1]

    prey[i] <- max(prey[i - 1] + dprey * dt, 0)
    predator[i] <- max(predator[i - 1] + dpredator * dt, 0)
  }

  data.frame(time = time, prey = prey, predator = predator)
}

params <- read.csv(param_path)

rows <- list()

for (i in seq_len(nrow(params))) {
  p <- params[i, ]
  sim <- simulate_predator_prey(p$prey0, p$predator0, p$alpha, p$beta, p$delta, p$gamma, p$time_end, p$dt)

  rows[[i]] <- data.frame(
    scenario = p$scenario,
    final_prey = tail(sim$prey, 1),
    final_predator = tail(sim$predator, 1),
    max_prey = max(sim$prey),
    max_predator = max(sim$predator),
    mean_prey = mean(sim$prey),
    mean_predator = mean(sim$predator)
  )
}

summary_df <- do.call(rbind, rows)
print(round(summary_df, 4))
